src/matrix/decomposition/lu.rs

```
@@ -299,6 +295,253 @@ impl<T> PartialPivLu<T> where T: Any + Float {
+// back substitution uses Any bound)
+impl<T> FullPivLu<T> where T: Any + Float {
+
+    /// Solves the linear system `Ax = b`.
+    ///
+    /// Here, `A` is the decomposed matrix satisfying
+    /// `PAQ = LU`. Note that this method is particularly
+    /// well suited to solving multiple such linear systems
+    /// involving the same `A` but different `b`.
+    ///
+    /// # Errors
+    ///
+    /// If the matrix is very ill-conditioned, the function
+    /// might fail to obtain the solution to the system.
+    ///
+    /// # Panics
+    ///
+    /// The right-hand side vector `b` must have compatible size.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[macro_use] extern crate rulinalg;
+    /// # use rulinalg::matrix::decomposition::FullPivLu;
+    /// # use rulinalg::matrix::Matrix;
+    /// # fn main() {
+    /// let x = Matrix::identity(4);
+    /// let lu = FullPivLu::decompose(x).unwrap();
+    /// let b = vector![3.0, 4.0, 2.0, 1.0];
+    /// let y = lu.solve(b)
+    ///           .expect("Matrix is invertible.");
+    /// assert_vector_eq!(y, vector![3.0, 4.0, 2.0, 1.0], comp = float);
+    /// # }
+    /// ```
+    pub fn solve(&self, b: Vector<T>) -> Result<Vector<T>, Error> {
+        assert!(b.size() == self.lu.rows(),
+            "Right-hand side vector must have compatible size.");
+
+        let b = lu_forward_substitution(&self.lu, &self.p * b);
+        back_substitution(&self.lu, b).map(|x| &self.q * x)
+    }
+
+    /// Computes the inverse of the matrix which this LUP decomposition
+    /// represents.
+    ///
+    /// # Errors
+    /// The inversion might fail if the matrix is very ill-conditioned.
+    /// The inversion fails if the matrix is not square, or is not invertible.
+    pub fn inverse(&self) -> Result<Matrix<T>, Error> {
+        let n = self.lu.rows();
+        let mut inv = Matrix::zeros(n, n);
+        let mut e = Vector::zeros(n);
+
+        if self.rank() != n {
```

Well, `solve` can fail on the `n`th iteration, in which case it would be preferable to fail early, so perhaps a cheaper check beforehand like the one you had would be a good idea. In that case, we could perhaps do even better thank rank: we could implement something like `is_invertible` which simply checks the value of the bottom right corner of the `lu` matrix. But this doesn't have to be implemented for this PR :-)